contributors/design-proposals/admission_control_extension.md

```diff
@@ -0,0 +1,507 @@
+OriginNamespaceLifecycle | lifecycle/admission.go | Controls accepting resources for namespaces.
+
+
+### Patterns
+
+In a study of all known admission controllers, the following patterns were seen most often:
+
+1. Defaulting on creation
+2. Synchronous validation on creation
+3. Synchronous validation on update - side-effect free
+
+Other patterns seen less frequently include:
+
+1. Defaulting on update
+2. Resolving / further specifying values on update (ImagePolicy)
+3. Creating resources in response to user action with the correct permission check (JenkinsBootstrapper)
+4. Policy decisions based on *who* is doing the action (JenkinsBootstrapper)
+5. Synchronous validation on update - with side effects (quota)
+
+While admission controllers can operate on all verbs, resources, and sub resource types, in practice they
+mostly deal with create and update on primary resources. Most sub resources are highly privileged operations
+and so are typically covered by policy. Other controllers like quota tend to be per apiserver and therefore
+are not required to be extensible.
+
+
+### Building enforcement
+
+In order to implement custom admission, an admin, integrator, or distribution of Kubernetes must compile their
+admission controller(s) into the Kubernetes `kube-apiserver` binary. As Kubernetes is intended to be a
+modular layered system this means core components must be upgraded to effect policy changes and only a fixed
+list of plugins can be used. It also prevents experimentation and prototyping of policy, or "quick fix"
+solutions applied on site. As we add additional APIs that are not hosted in the main binary (either as third
+party resources or API extension servers), these APIs have many of the same security and policy needs that
+the core resources do, but must compile in their own subsets of admission.
+
+Further, distributions of Kubernetes like OpenShift that wish to offer complete solutions (such as OpenShift's
+multi-tenancy model) have no mechanism for running on top of Kubernetes without recompilation of the core or
+for extending the core with additional policy. This prevents the formation of an open ecosystem for tools
+*around* Kubernetes, forcing all changes to policy to go through the Kubernetes codebase review gate (when
+such review is unnecessary or disruptive to Kubernetes itself).
+
+
+## Design
+
+It should be possible to perform holistic policy enforcement in Kubernetes without the recompilation of the
+core project as plugins that can be added and removed to a stock Kubernetes release. That extension
+of admission control should leverage similar mechanisms to our existing controller frameworks where possible
```

Comment:
What does it mean by "leverage similar mechanism to our existing controller frameworks", "otherwise be performant and reliable"?

This is confusing..